datafusion/functions-array/src/kernels.rs

```diff
@@ -0,0 +1,254 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! implementation kernels for array functions
+
+use arrow::array::{
+    Array, ArrayRef, BooleanArray, Float32Array, Float64Array, GenericListArray,
+    Int16Array, Int32Array, Int64Array, Int8Array, LargeStringArray, OffsetSizeTrait,
+    StringArray, UInt16Array, UInt32Array, UInt64Array, UInt8Array,
+};
+use arrow::datatypes::DataType;
+use datafusion_common::cast::{as_large_list_array, as_list_array, as_string_array};
+use datafusion_common::{exec_err, DataFusionError};
+use std::any::type_name;
+use std::sync::Arc;
+
+macro_rules! downcast_arg {
```

Comment:
we may want to move into utils,`math_expressions.rs` has very similar macros, I'll do it in following PR

Reply:
Thanks @comphead  -- once this PR is approved and merged, that would be great

I suspect as we go about porting over the old to new code, we will find other examples that can be cleaned up (some of the functions were implemented very early in DataFusion's life)